internal/controllers/bundledeployment/bundledeployment.go

```diff
@@ -198,53 +249,79 @@ func (c *controller) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 	bd.Status.ObservedGeneration = bd.Generation
 
-	bundle, allBundles, err := util.ReconcileDesiredBundle(ctx, c.cl, bd)
+	// handle finalizers
+	finalizerBundleDelpoyment := bd.DeepCopy()
+	finalizerResult, err := c.finalizers.Finalize(ctx, finalizerBundleDelpoyment)
+
 	if err != nil {
+		bd.Status.ResolvedSource = nil
+		bd.Status.ContentURL = ""
 		meta.SetStatusCondition(&bd.Status.Conditions, metav1.Condition{
-			Type:    rukpakv1alpha1.TypeHasValidBundle,
+			Type:    rukpakv1alpha2.TypeUnpacked,
 			Status:  metav1.ConditionUnknown,
-			Reason:  rukpakv1alpha1.ReasonReconcileFailed,
+			Reason:  rukpakv1alpha2.ReasonProcessingFinalizerFailed,
 			Message: err.Error(),
 		})
 		return ctrl.Result{}, err
 	}
-	if bundle.Status.Phase != rukpakv1alpha1.PhaseUnpacked {
-		reason := rukpakv1alpha1.ReasonUnpackPending
-		status := metav1.ConditionTrue
-		message := fmt.Sprintf("Waiting for the %s Bundle to be unpacked", bundle.GetName())
-		if bundle.Status.Phase == rukpakv1alpha1.PhaseFailing {
-			reason = rukpakv1alpha1.ReasonUnpackFailed
-			status = metav1.ConditionFalse
-			message = fmt.Sprintf("Failed to unpack the %s Bundle", bundle.GetName())
-			if c := meta.FindStatusCondition(bundle.Status.Conditions, rukpakv1alpha1.TypeUnpacked); c != nil {
-				message = fmt.Sprintf("%s: %s", message, c.Message)
-			}
-		}
-		meta.SetStatusCondition(&bd.Status.Conditions, metav1.Condition{
-			Type:    rukpakv1alpha1.TypeHasValidBundle,
-			Status:  status,
-			Reason:  reason,
-			Message: message,
-		})
+
+	if finalizerResult.Updated {
+		// The only thing outside the status that should ever change when handling finalizers
+		// is the list of finalizers in the object's metadata. In particular, we'd expect
+		// finalizers to be added or removed.
+		bd.ObjectMeta.Finalizers = finalizerBundleDelpoyment.ObjectMeta.Finalizers
+	}
+	if finalizerResult.StatusUpdated {
+		bd.Status = finalizerBundleDelpoyment.Status
+	}
+	if finalizerResult.Updated || finalizerResult.StatusUpdated || !bd.GetDeletionTimestamp().IsZero() {
 		return ctrl.Result{}, nil
 	}
 
-	meta.SetStatusCondition(&bd.Status.Conditions, metav1.Condition{
-		Type:    rukpakv1alpha1.TypeHasValidBundle,
-		Status:  metav1.ConditionTrue,
-		Reason:  rukpakv1alpha1.ReasonUnpackSuccessful,
-		Message: fmt.Sprintf("Successfully unpacked the %s Bundle", bundle.GetName()),
-	})
+	unpackResult, err := c.unpacker.Unpack(ctx, bd)
+	if err != nil {
+		return ctrl.Result{}, updateStatusUnpackFailing(&bd.Status, fmt.Errorf("source bundle content: %v", err))
+	}
+
+	switch unpackResult.State {
+	case unpackersource.StatePending:
+		updateStatusUnpackPending(&bd.Status, unpackResult)
+		// There must a limit to number of retries if status is stuck at
+		// unpack pending.
+		// Forcefully requing here, to ensure that the next reconcile
+		// is triggered.
+		// TODO: Caliberate the requeue interval if needed.
```

Comment:
I think the expectation here is that the controller will receive an event when the underlying unpacking process moves beyond Pending, so I'm wondering if there's a need to use `RequeueAfter`?

The Pending state is used:
- by the 'image' unpacker which uses a pod (and we get pod events when the pod moves on from pending)
- by the 'upload' source. When the upload manager receives an upload it directly updates the status of the BD, which triggers a reconcile.